### PR TITLE
Fix persistence allocations #5505.

### PR DESCRIPTION
**Pull request by F0b0s**
Fixes #

## Changes

As discussed in #5505 closure object creation for `WithCircuitBreaker<TResult>` was fixed. Also all local functions like `ProcessResults`, `ExecuteBatch`, `Resequence` was moved into class method because every call ends with closure object creation. 
For `AtomicWrite` linq methods was removed usage because it ends in memory allocation.
It shows good consumed memory performance for write journal critical path:
`JournalWriteBenchmarks`  - 4% lower.
`AtomicWriteBenchmark` - 25% lower

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).

### Latest `dev` Benchmarks 

BenchmarkDotNet=v0.13.2, OS=macOS Monterey 12.5.1 (21G83) [Darwin 21.6.0]
Apple M1 2.40GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.101
  [Host] : .NET 7.0.1 (7.0.122.56804), X64 RyuJIT SSE4.2 DEBUG

JournalWriteBenchmarks

|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |   StdDev |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|---------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    789.1 us |  47.17 us | 133.8 us |         - |         - |   163.88 KB |
| WriteToPersistence |               10 |           100 |  1,868.0 us |  50.21 us | 145.7 us |         - |         - |  1582.23 KB |
| WriteToPersistence |              100 |           100 | 24,981.2 us | 492.14 us | 657.0 us | 2000.0000 | 1000.0000 | 15686.82 KB |

AtomicWriteBenchmark

|      Method | WriteMsgCount |     Mean |     Error |    StdDev |      Gen0 | Allocated |
|------------ |-------------- |---------:|----------:|----------:|----------:|----------:|
| Constructor |        100000 | 9.324 ms | 0.1661 ms | 0.2434 ms | 1000.0000 |   8.39 MB |

### This PR's Benchmarks

InJournalWriteBenchmarks

|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |   StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|---------:|------------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    785.2 us |  50.09 us | 143.7 us |    756.2 us |         - |         - |   156.91 KB |
| WriteToPersistence |               10 |           100 |  1,811.6 us |  54.72 us | 157.9 us |  1,765.2 us |         - |         - |  1518.79 KB |
| WriteToPersistence |              100 |           100 | 23,806.4 us | 463.65 us | 694.0 us | 23,761.8 us | 2000.0000 | 1000.0000 | 14953.15 KB |

AtomicWriteBenchmark

|      Method | WriteMsgCount |     Mean |     Error |    StdDev |      Gen0 | Allocated |
|------------ |-------------- |---------:|----------:|----------:|----------:|----------:|
| Constructor |        100000 | 8.652 ms | 0.1305 ms | 0.2031 ms | 1000.0000 |    6.1 MB |

### CODE REVIEW

**Comment by F0b0s** [comment]
I can try to rewrite it with local functions as it was earlier, up to you @Aaronontheweb

**Comment by Aaronontheweb** [comment]
> I can try to rewrite it with local functions as it was earlier, up to you @Aaronontheweb

I'll take a look at this as soon as I can!

BTW, I recommend that you join the Akka.NET Discord and join the #contributors room  - we kick around ideas in there every day usually. https://discord.gg/GSCfPwhbWP

File: src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs [hunk 1]
```diff
@@ -223,8 +223,8 @@ async Task ProcessDelete()
             {
                 try
                 {
-                    await _breaker.WithCircuitBreaker(() =>
-                        DeleteMessagesToAsync(message.PersistenceId, message.ToSequenceNr));
+                    await _breaker.WithCircuitBreaker((message, awj: this), state =>
+                        state.awj.DeleteMessagesToAsync(state.message.PersistenceId, state.message.ToSequenceNr));
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs:227]: lgtm

File: src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs [hunk 2]
```diff
@@ -272,8 +272,8 @@ void CompleteHighSeqNo(long highSeqNo)
                 
                 try
                 {
-                    var highSequenceNr = await _breaker.WithCircuitBreaker(() =>
-                        ReadHighestSequenceNrAsync(message.PersistenceId, readHighestSequenceNrFrom));
+                    var highSequenceNr = await _breaker.WithCircuitBreaker((message, readHighestSequenceNrFrom, awj: this), state =>
+                        state.awj.ReadHighestSequenceNrAsync(state.message.PersistenceId, state.readHighestSequenceNrFrom));
```
Comment (Aaronontheweb) [inline src/core/Akka.Persistence/Journal/AsyncWriteJournal.cs:276]: lgtm

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Both sets of multi-node specs are failing using Akka.Persistence + Cluster.Sharding on this PR, which makes me think this change has some unintended side effects. Needs to be investigated and resolved before we can merge.

**Comment by F0b0s** [comment]
Ok, will take a look.

**Comment by F0b0s** [comment]
Found an error in my PR,
source code was like:
`                var prepared = PreparePersistentBatch(message.Messages).ToArray();`
i changed it to 
`                var prepared = PreparePersistentBatch(message.Messages);`
and as result of `PreparePersistentBatch` is `IEnumerable`, ToArray play role of materializer, i return it back and all tests are passed.

@Aaronontheweb can you take a look again.

Current benchmark values:
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |   StdDev |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|---------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    861.3 us |  54.74 us | 157.9 us |         - |         - |   159.08 KB |
| WriteToPersistence |               10 |           100 |  1,754.1 us |  37.88 us | 106.8 us |         - |         - |  1519.92 KB |
| WriteToPersistence |              100 |           100 | 24,152.2 us | 480.19 us | 571.6 us | 2000.0000 | 1000.0000 | 15190.59 KB |

**Comment by F0b0s** [comment]
And i rewrited linq calls in `WriteJournalBase` and memory allocation decreased on 10% percent in `JournalWriteBenchmarks`

|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |   StdDev |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|---------:|----------:|----------:|------------:|
| WriteToPersistence |                1 |           100 |    805.2 us |  49.43 us | 141.8 us |         - |         - |   149.81 KB |
| WriteToPersistence |               10 |           100 |  1,699.2 us |  43.27 us | 124.1 us |         - |         - |  1423.88 KB |
| WriteToPersistence |              100 |           100 | 22,949.3 us | 457.68 us | 685.0 us | 2000.0000 | 1000.0000 | 14178.66 KB |

**Comment by Aaronontheweb** [comment]
I'll re-check the numbers of my machine and give you some feedback

**Comment by Aaronontheweb** [comment]
On my Ryzen machine.

## `dev`

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2486/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-JKMUOM : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **502.9 μs** |  **47.25 μs** |   **137.1 μs** |         **-** |         **-** |   **164.11 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,881.5 μs** | **194.93 μs** |   **565.5 μs** |         **-** |         **-** |  **1578.97 KB** |
| **WriteToPersistence** |              **100** |           **100** | **31,122.9 μs** | **690.19 μs** | **1,854.1 μs** | **2000.0000** | **1000.0000** | **15608.03 KB** |


## This PR

``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2486/21H2/November2021Update)
AMD Ryzen 7 1700, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  Job-DUCEDZ : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2

InvocationCount=1  UnrollFactor=1  

```
|             Method | PersistentActors | WriteMsgCount |        Mean |     Error |     StdDev |      Median |      Gen0 |      Gen1 |   Allocated |
|------------------- |----------------- |-------------- |------------:|----------:|-----------:|------------:|----------:|----------:|------------:|
| **WriteToPersistence** |                **1** |           **100** |    **464.1 μs** |  **37.16 μs** |   **109.0 μs** |    **479.8 μs** |         **-** |         **-** |   **148.86 KB** |
| **WriteToPersistence** |               **10** |           **100** |  **1,610.2 μs** | **190.79 μs** |   **559.6 μs** |  **1,525.0 μs** |         **-** |         **-** |  **1422.73 KB** |
| **WriteToPersistence** |              **100** |           **100** | **29,956.3 μs** | **852.72 μs** | **2,348.6 μs** | **29,288.3 μs** | **2000.0000** | **1000.0000** | **14044.73 KB** |

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**